### PR TITLE
ODF units harmonization

### PR DESCRIPTION
**Pull request by JessyBarrette**
# Issue
ERDDAP review for each source file the units recorded for each variables and source files it ingest to be sure that all of them are consistent. This is a good thing! But as you're ingesting multiple years of data, those units terms can vary over time, as an example depth values in  `m`, `metre`, `meter` and `metres` won't be considered the same by ERDDAP. 

# Method for correcting
This present PR tries to handle this issue by comparing the units to the matching vocabulary and term within the Vocabulary CSV list file. For each variables retrieved in an ODF, the tools review the matching terms associated within the vocabularies listed in the config.json file and by considering the vocabularies priority order. 

To be considered, a variable should match the the vocabulary, variable name and matching one of the units listed under the expected_units column, if any matching is found, the very first unit within the list of units is used instead of the original one (which remain saved under the original_UNITS attribute).
## Special Case: units=None
If the term None is available within the `expected_units` listed, the tool assume that the absence of unit or the terms `(none)`, `None `and `none` are considered acceptable for this specific variable. If so, the first term listed within the `expected_units` will still be used. This is particularly useful will dimensionless units (ex: practical salinity), but should be handle carefully usually not dimentionless data. 

# Example
An ODF files with the variable `TEMP_01 [IPTS-68, deg C]` will match the term:
| Vocabulary| Name| expected_units|
|---|---|---|
| GF3 |TEMP |"degree Celsius&#124;degree c&#124;degrees C&#124;IPTS-68, deg C&#124;ITS-68, deg C"|

Based on that match, the unit will be replaced by the term `degree Celsius`, If available, like in this example, the scale terms (`IPTS-48`, `IPTS-68`, `ITS-90` or `PSS-78`) transferred to a new attribute `scale`. This will then result associating the following attributes to the TEMP_01 variable:
``` python
{TEMP_01: {
    units: "degree_celsius",
    scale: "IPTS-68".
    ...
    }
}
```

# Matching units available and test on historical data
The CSV vocabulary contains as of now all possible unit terms available within the ~2500 odf test files.  

Some historical data from BIO do not contain any units. Those specific files will need to be reviewed with the organizations and see would be the best to handle them.

Some issues were detected within the historical data which has some minor format changes over the years.  Corrections to handle those different issues will be presented into another PR.

### CODE REVIEW

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py [hunk 1]
```diff
@@ -316,6 +353,33 @@ def _find_previous_key(key_list, key):
     return metadata
 
 
+def parse_odf_code_variable(var_name):
```
Comment (n-a-t-e) [inline projects/odf_transform/odf_transform/odf_parser/odf_parser.py:356]: can you add docstring to this one?

File: projects/odf_transform/odf_transform/odf_to_netcdf.py [hunk 2]
```diff
@@ -24,13 +24,20 @@ def read_config(config_file):
         config = json.load(fid)
 
         # Read Vocabulary file
+        config["vocabulary"] = {}
         for vocab_file in config["vocabularyFileList"]:
-            config.update({"vocabulary": {}})
-            with open(vocab_file) as fid:
-                vocab = json.load(fid)
-            config["vocabulary"].update(vocab)
+            if vocab_file.endswith('json'):
```
Comment (n-a-t-e) [inline projects/odf_transform/odf_transform/odf_to_netcdf.py:29]: can we remove json support if we're swtiching to CSV, just to simplify

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py [hunk 3]
```diff
@@ -207,54 +207,91 @@ def _find_previous_key(key_list, key):
...
+            flag_column = True
+        # Make sure that the flag column relate to something
+        if flag_column and parameter_code[0] not in ['QCFF', 'FFFF'] and flag_dict[odf_parameter_code] not in metadata:
+            warnings.warn(odf_parameter_code + ' flag is refering to' + \
+                          flag_dict[odf_parameter_code] + ' which is not available as variable',
+                          UserWarning)
+
+        # Loop through each organisations and find the matching parameter_code within the vocabulary
+        standardized_variables = False
+        if type(vocabulary) is pd.DataFrame and not flag_column and var not in ['SYTM_01']:
```
Comment (n-a-t-e) [inline projects/odf_transform/odf_transform/odf_parser/odf_parser.py:246]: which other type are you expecting `vocabulary` could be?

Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:246]: yeah we'll move away from the json ok I'll add those changes

File: projects/odf_transform/odf_transform/odf_to_netcdf.py (hunk 2 above)
Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_to_netcdf.py:29]: let's remove the json option

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py [hunk 4]
```diff
@@ -207,54 +207,91 @@ def _find_previous_key(key_list, key):
...
-            metadata[var]['parameter_code'] = parameter_code[0]
-            if len(parameter_code) == 2 and parameter_code[1] != '':
-                metadata[var]['parameter_code_number'] = int(parameter_code[1])
+
+    flag_dict = {}
+    for var in metadata.keys():
+        # Retrieve ODF CODE and Associated number
+        odf_parameter_code = metadata[var][parameter_code_att]
+        # Separate the parameter_code from the number at the end of the variable
+        parameter_code = odf_parameter_code.rsplit('_', 1)
```
Comment (n-a-t-e) [inline projects/odf_transform/odf_transform/odf_parser/odf_parser.py:216]: could be more clear if you could call these variables something like `parameter_code_parts` (the split string) `parameter_code` (just the text code) and `parameter_code_number` (the sequence number) instead of using `parameter_code[0]`

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py (hunk 1 above)
Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:356]: Docstring was added!

**Review by n-a-t-e — COMMENTED** [review]
odf_parser.py is getting pretty long, would be good to split out just the code that creates the dictionary into it's own file at some point (doesnt have to be this PR)

File: projects/odf_transform/odf_transform/odf_to_netcdf.py (hunk 2 above)
Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_to_netcdf.py]: I'll also make it only expect only a single vocabulary csv file

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py (hunk 3 above)
Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py]: it was replace by `vocabulary is not None`

File: cioos_data_transform/cioos_data_transform/utils/xarray_methods.py [hunk 5]
```diff
@@ -232,23 +232,38 @@ def define_index_dimensions(ds):
 
 def add_variable_attributes(ds,
                             review_attributes=None,
-                            scales=r'IPTS\-*48|IPTS\-*68|ITS\-*90|PSS\-*78|practical\ssalinity|psu'):
+                            overwrite=True):
+
+    def _get_scale(attributes):
```
Comment (n-a-t-e) [inline cioos_data_transform/cioos_data_transform/utils/xarray_methods.py:237]: `attributes` is unused

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py [hunk 6]
```diff
@@ -316,6 +353,38 @@ def _find_previous_key(key_list, key):
...
+    if len(var_list) > 1 and var_list[1] not in ['']:
+        var_dict['index'] = int(var_list[1])
+        var_dict['standardized_name'] += '_{0:02.0f}'.format(var_dict['index'])
+    elif len(var_list) > 1 and var_list[1] == '':
+        var_dict['standardized_name'] += '_'
+
+    return var_dict
+
+
+def standardize_odf_units(unit_string, escape=False):
```
Comment (n-a-t-e) [inline projects/odf_transform/odf_transform/odf_parser/odf_parser.py:356]: `escape` is unused

File: projects/odf_transform/odf_transform/vocabulary/CIOOS_Vocabulary_List - Complete_Vocabulary.csv [hunk 7]
```diff
@@ -0,0 +1,1480 @@
+Vocabulary,name,Users,MEDS:STATUS,MEDS:CATEGORY_MEANING_E,MEDS:CATEGORY_MEANING_F,MEDS:UNITS_DESCRIPTOR_E,MEDS:UNITS_DESCRIPTOR_F,MEDS:CF_CODE,MEDS:NETCDF_LONG_NAME,MEDS:WMO_CODE_TABLE_ID,MEDS:CONVENTION,BIO:units,BIO:P06 urn,BIO:P06 name,BIO:P01 urn,BIO:P01 name,BIO:CF_CODE,IML:definition,IML:units,long_name,expected_units,scale,standard_name,SDN:P07::id,SDN:P01::id,SDN:P01::urn,SDN:P01::prefLabel,sdn_parameter_urn,sdn_parameter_name,SDN:P06::id,SDN:P06::urn,SDN:P06::prefLabel,sdn_uom_urn,sdn_uom_name,definition,convention,scale_factor,GF3_name,Alternative units,SDN:P01::Alternative Name
```
Comment (n-a-t-e) [inline projects/odf_transform/odf_transform/vocabulary/CIOOS_Vocabulary_List - Complete_Vocabulary.csv:1]: looks like there's a mix of and `|` and `,` separators for `expected_units` - is it supposed to be like that? eg `"kg/m^3|kilogram per metre cube|sigma-theta, kg/m^3|sigma-theta, Kg/m^3"`

Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/vocabulary/CIOOS_Vocabulary_List - Complete_Vocabulary.csv:1]: no the unit has that comma only '|' is a sperator
if comma exist in the string the expected unit needs to be double quoted

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py (hunk 6 above)
Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:356]: yes let's get rid of that

**Review by n-a-t-e — APPROVED** [review]

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py [hunk 8]
```diff
@@ -39,32 +30,24 @@ def read(filename,
...
+    header_end = '-- DATA --'
+    data_delimiter = r'\s+'
+    quotechar = '\''
+    parameter_section = 'PARAMETER_HEADER'
+    variable_type = 'TYPE'
+    null_value = 'NULL_VALUE'
+    section_items_minimum_whitespaces = 2
+    original_prefix_var_attribute = 'original_'
 
     def _convert_to_number(value):
```
Comment (n-a-t-e) [inline projects/odf_transform/odf_transform/odf_parser/odf_parser.py:52]: not part of this PR but looks like this function has some issues, eg 
`_convert_to_number(5.4)` = 5

File: cioos_data_transform/cioos_data_transform/utils/xarray_methods.py (hunk 5 above)
Comment (JessyBarrette) [reply cioos_data_transform/cioos_data_transform/utils/xarray_methods.py]: This was removed!

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py (hunk 4 above)
Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:209]: That's a good idea!

Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:209]: There's actually a new function to handle the split of standardization of those codes. It may be better to use that instead.

Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py]: This was replaced by parse_odf_code_variable function

File: projects/odf_transform/odf_transform/odf_parser/odf_parser.py (hunk 8 above)
Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:52]: oh yeah very good point! Float and then int!

Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:52]: This was corrected!

Comment (JessyBarrette) [reply projects/odf_transform/odf_transform/odf_parser/odf_parser.py:52]: that was corrected but should have been actually need to be corrected in next PR
